{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "manual-screening",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b73e8ab8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# to FASTA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "equal-confidentiality",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted 2 records\n"
     ]
    }
   ],
   "source": [
    "sequences = []\n",
    "\n",
    "with open(r'C:\\Users\\e_mulciber\\Desktop\\MitoClub\\devil-worm\\Body\\1Raw\\mtDNA - mephisto mtDNA project.gb', \"r\") as input_handle:\n",
    "    with open(r'C:\\Users\\e_mulciber\\Desktop\\MitoClub\\devil-worm\\Body\\1Raw\\halicephalobus.fasta', \"w\") as output_handle:\n",
    "        for read in SeqIO.parse(input_handle, \"genbank\"):\n",
    "            if 'halicephalobus' in read.description or 'Halicephalobus' in read.annotations['organism']:\n",
    "                sequences.append(read)\n",
    "        count = SeqIO.write(sequences, output_handle, \"fasta\")\n",
    "        \n",
    "print(\"Converted %i records\" % count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "addc99ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# to GENBANK derive Nematoda dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c6386e8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted 190 records\n"
     ]
    }
   ],
   "source": [
    "sequences = []\n",
    "species = set()\n",
    "\n",
    "with open('/home/emulciber/MitoClub/devil-worm/Body/2Derived/nematoda.gbff', \"w\") as output_handle:\n",
    "    with open('/home/emulciber/MitoClub/mtDnaInvertebrate/Body/1Raw/mitochondrion.genomic.gbff', \"r\") as input_handle:\n",
    "        for read in SeqIO.parse(input_handle, \"genbank\"):\n",
    "            if 'Nematoda' in read.annotations['taxonomy']:\n",
    "                sequences.append(read)\n",
    "                species.add(read.annotations['organism'])\n",
    "    with open('/home/emulciber/MitoClub/devil-worm/Body/1Raw/mtDNA - mephisto mtDNA project.gb', \"r\") as input_handle:\n",
    "        for read in SeqIO.parse(input_handle, \"genbank\"):\n",
    "            if read.annotations['organism'] not in species and read.annotations['organism'] != 'Homo sapiens' and read.annotations['organism'] != 'Drosophila melanogaster':\n",
    "                sequences.append(read)    \n",
    "    with open('/home/emulciber/MitoClub/devil-worm/Body/1Raw/plectus_aquatilis.gb', \"r\") as input_handle:\n",
    "        for read in SeqIO.parse(input_handle, \"genbank\"):\n",
    "            sequences.append(read)\n",
    "        \n",
    "    count = SeqIO.write(sequences, output_handle, \"genbank\")\n",
    "        \n",
    "print(\"Converted %i records\" % count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f2d4f04",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Проверка"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5e4b5306",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bursaphelenchus xylophilus\n",
      "Hypodontus macropi\n",
      "Macropicola ocydromi\n",
      "Protostrongylus rufescens\n",
      "Toxascaris leonina\n",
      "Rhigonema thysanophora\n",
      "Nematodirus spathiger\n",
      "Nematodirus oiratianus\n",
      "Parafilaroides normani\n",
      "Oesophagostomum asperum\n",
      "Oesophagostomum columbianum\n",
      "Contracaecum osculatum\n",
      "Parascaris univalens\n",
      "Philometroides sanguineus\n",
      "Aphelenchoides besseyi\n",
      "Caenorhabditis tropicalis\n",
      "Anisakis berlandi\n",
      "Trichinella pseudospiralis\n",
      "Trichinella britovi\n",
      "Trichinella murrelli\n",
      "Trichinella nativa\n",
      "Trichinella nelsoni\n",
      "Trichinella papuae\n",
      "Trichinella zimbabwensis\n",
      "Meloidogyne incognita\n",
      "Gongylonema pulchrum\n",
      "Triodontophorus brevicauda\n",
      "Strongylus equinus\n",
      "Oxyuris equi\n",
      "Uncinaria sanguinis\n",
      "Meloidogyne arenaria\n",
      "Meloidogyne enterolobii\n",
      "Meloidogyne javanica\n",
      "Pseudoterranova azarasi\n",
      "Litoditis aff. marina PmI\n",
      "Litoditis aff. marina PmII\n",
      "Litoditis aff. marina PmIII\n",
      "Litoditis aff. marina PmIV\n",
      "Gnathostoma spinigerum\n",
      "Strongyloides venezuelensis\n",
      "Strongyloides stercoralis\n",
      "Strongyloides ratti\n",
      "Trichuris muris\n",
      "Parastrongyloides trichosuri\n",
      "Strongyloides papillosus\n",
      "Allodiplogaster sudhausi\n",
      "Haemonchus placei\n",
      "Passalurus ambiguus\n",
      "Heterakis beramporia\n",
      "Heterakis gallinarum\n",
      "Dirofilaria repens\n",
      "Thelazia callipaeda\n",
      "Trichuris discolor\n",
      "Trichuris ovis\n",
      "Chabertia erschowi\n",
      "Angiostrongylus malaysiensis\n",
      "Angiostrongylus vasorum\n",
      "Dirofilaria sp. 'hongkongensis'\n",
      "Triodontophorus serratus\n",
      "Triodontophorus nipponicus\n",
      "Gnathostoma doloresi\n",
      "Cylicocyclus nassatus\n",
      "Longidorus vineacola\n",
      "Paralongidorus litoralis\n",
      "Xiphinema rivesi\n",
      "Xiphinema pachtaicum\n",
      "Gnathostoma nipponicum\n",
      "Ancylostoma tubaeforme\n",
      "Anisakis pegreffii\n",
      "Diploscapter coronatus\n",
      "Cyathostomum catinatum\n",
      "Cylicostephanus minutus\n",
      "Poteriostomum imparidentatum\n",
      "Ancylostoma ceylanicum\n",
      "Caenorhabditis remanei\n",
      "Caenorhabditis brenneri\n",
      "Caenorhabditis plicata\n",
      "Caenorhabditis angaria\n",
      "Caenorhabditis doughertyi\n",
      "Caenorhabditis virilis\n",
      "Caenorhabditis wallacei\n",
      "Caenorhabditis nouraguensis\n",
      "Caenorhabditis macrosperma\n",
      "Caenorhabditis afra\n",
      "Caenorhabditis castelli\n",
      "Onchocerca volvulus\n",
      "Marshallagia marshalli\n",
      "Parascaris equorum\n",
      "Ascaris ovis\n",
      "Ortleppascaris sinensis\n",
      "Pseudoterranova bulbosa\n",
      "Pseudoterranova cattani\n",
      "Pseudoterranova decipiens sensu lato Germany\n",
      "Pseudoterranova krabbei\n",
      "Contracaecum ogmorhini sensu lato Australia\n",
      "Contracaecum ogmorhini sensu lato South Africa\n",
      "Contracaecum ogmorhini sensu lato Pacific Canada\n",
      "Nippostrongylus brasiliensis\n",
      "Camallanus cotti\n",
      "Cyathostomum pateratum\n",
      "Cylicocyclus radiatus\n",
      "Onchocerca ochengi\n",
      "Heterorhabditis indica\n",
      "Cylicodontophorus bicoronatus\n",
      "Cylicocyclus auriculatus\n",
      "Setaria labiatopapillosa\n",
      "Heterakis dispar\n",
      "Steinernema abbasi\n",
      "Angiostrongylus mackerrasae\n",
      "Cylicocyclus ashworthi\n",
      "Dictyocaulus viviparus\n",
      "Dictyocaulus eckerti\n",
      "Caenorhabditis elegans\n",
      "Eucoleus annulatus\n",
      "Brugia malayi\n",
      "Enterobius vermicularis\n",
      "Onchocerca lupi\n",
      "Meloidogyne graminicola\n",
      "Agamermis sp. BH-2006\n",
      "Heterorhabditis bacteriophora\n",
      "Romanomermis culicivorax\n",
      "Romanomermis nielseni\n",
      "Romanomermis iyengari\n",
      "Hexamermis agrotis\n",
      "Trichinella spiralis\n",
      "Caenorhabditis briggsae\n",
      "Toxocara malaysiensis\n",
      "Toxocara canis\n",
      "Toxocara cati\n",
      "Ancylostoma duodenale\n",
      "Haemonchus contortus\n",
      "Bunostomum phlebotomum\n",
      "Ancylostoma caninum\n",
      "Angiostrongylus cantonensis\n",
      "Angiostrongylus costaricensis\n",
      "Trichostrongylus vitrinus\n",
      "Metastrongylus pudendotectus\n",
      "Metastrongylus salmi\n",
      "Oesophagostomum dentatum\n",
      "Syngamus trachea\n",
      "Trichostrongylus axei\n",
      "Teladorsagia circumcincta\n",
      "Chabertia ovina\n",
      "Mecistocirrus digitatus\n",
      "Strongylus vulgaris\n",
      "Oesophagostomum quadrispinulatum\n",
      "Setaria digitata\n",
      "Cooperia oncophora\n",
      "Contracaecum rudolphii B Bullini et al., 1986\n",
      "Pristionchus pacificus\n",
      "Baylisascaris transfuga\n",
      "Baylisascaris schroederi\n",
      "Necator americanus\n",
      "Dracunculus medinensis\n",
      "Heliconema longissimum\n",
      "Cucullanus robustus\n",
      "Wellcomia siamensis\n",
      "Ascaris lumbricoides\n",
      "Baylisascaris procyonis\n",
      "Trichuris trichiura\n",
      "Trichuris suis\n",
      "Dirofilaria immitis\n",
      "Steinernema carpocapsae\n",
      "Xiphinema americanum\n",
      "Ascaris suum\n",
      "Anisakis simplex\n",
      "Thaumamermis cosgrovei\n",
      "Strelkovimermis spiculatus\n",
      "Chandlerella quiscali\n",
      "Onchocerca flexuosa\n",
      "Wuchereria bancrofti\n",
      "Acanthocheilonema viteae\n",
      "Bunostomum trigonocephalum\n",
      "Aelurostrongylus abstrusus\n",
      "Bursaphelenchus mucronatus\n",
      "Spirocerca lupi\n",
      "Ascaridia galli\n",
      "Ascaridia columbae\n",
      "Cylicocyclus insigne\n",
      "Acrobeloides varius\n",
      "Caenorhabditis nigoni\n",
      "Diploscapter pachys (synthetic DNA construct)\n",
      "Halicephalobus mephisto\n",
      "Loa loa\n",
      "Panagrellus redivivus\n",
      "Steinernema glaseri\n",
      "Steinernema litorale\n",
      "Halicephalobus gingivalis\n",
      "Halicephalobus sp. NKZ332\n",
      "Plectus aquatilis\n"
     ]
    }
   ],
   "source": [
    "seq = SeqIO.parse('/home/emulciber/MitoClub/devil-worm/Body/2Derived/nematoda.gbff', 'genbank')\n",
    "for sequence in seq:\n",
    "    print(sequence.annotations['organism'])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
